core/src/main/java/brooklyn/location/access/PortForwardManagerImpl.java

```diff
@@ -181,6 +188,20 @@ protected void associateImpl(String publicIpId, HostAndPort publicEndpoint, Loca
         onChanged();
     }
 
+    private void emitAssociationCreatedEvent(String publicIpId, HostAndPort publicEndpoint, Location location, int privatePort) {
+        AssociationMetadata metadata = new AssociationMetadata(publicIpId, publicEndpoint, location, privatePort);
+        for (Map.Entry<AssociationListener, Predicate<? super AssociationMetadata>> entry : associationListeners.entrySet()) {
+            if (entry.getValue().apply(metadata)) {
+                try {
+                    entry.getKey().onAssociationCreated(metadata);
+                } catch (Exception e) {
+                    Exceptions.propagateIfFatal(e);
+                    log.warn("Exception thrown when emitting association creation event {}: {}", metadata, e);
```

Comment:
This will just give the toString of the exception, rather than the stacktrace - it would make debugging problems very hard. For example, it might tell you there was an NPE without ever telling you which line that null pointer happened on. Suggest changing it to:

    log.warn("Exception thrown when emitting association creation event "+metadata, e);

Same for log message in `emitAssociationDeletedEvent`.